### PR TITLE
Add Q&A search on contact message

### PR DESCRIPTION
**Pull request by notnownikki**
The Q&A endpoint on wpcom matches user's initial messages to support against a database of frequently asked questions, as a final automated search in an attempt to help them before they start chats or open tickets.

This PR adds as-you-type searching using the Q&A API, on the contact form.

### CODE REVIEW

**Comment by notnownikki** [comment]
There's a test question live in the database, using a single keyword that users are (probably) very unlikely to search for!

If you go to the contact form and ask for help with "quirkafleeg" (or "quirkafleeging", or "I don't know how to use my quirkafleegs!"), you can see this in action.

**Comment by notnownikki** [comment]
How to test:

1) Click the help icon in the lower left corner
2) Scroll to the bottom and click the "Contact us" button
3) In either the message or subject fields on the contact form, use the keyword "quirkafleeg"

Results should look like this:

quirk

**Comment by johngodley** [comment]
Works great with quirkafleeg!

The results display seems fine to me, but I don't know if it fits in with the Calypso style - it might be worth having a designer give a thumbs up for that.

Having several responses ready for quirkafleeg would help too, just so we can see what it looks like with more than one response.

File: client/me/help/help-contact-form/index.jsx [hunk 1]
```diff
@@ -99,6 +114,13 @@ export const HelpContactForm = React.createClass( {
 		}
 	},
 
+	doQandASearch() {
+		const query = this.state.subject + ' ' + this.state.message;
+		wpcom.getQandA( query, ( error, data ) => {
+			this.setState( { qanda: data } );
```
Comment (dmsnell) [inline client/me/help/help-contact-form/index.jsx]: I guess this can never ever fail for any reason?

Comment (notnownikki) [inline client/me/help/help-contact-form/index.jsx]: Gah, missed this! Yes, I'll add error handling in there.

**Comment by dmsnell** [comment]
@notnownikki are you familiar with the Promise-passing style of wpcom functions? if you don't pass a callback you can chain `then`s and `catch`s

File: client/me/help/help-contact-form/index.jsx [hunk 2]
```diff
@@ -99,6 +114,13 @@ export const HelpContactForm = React.createClass( {
 		}
 	},
 
+	doQandASearch() {
+		const query = this.state.subject + ' ' + this.state.message;
+		wpcom.getQandA( query, ( error, data ) => {
```
Comment (notnownikki) [inline client/me/help/help-contact-form/index.jsx]: So I added a condition that blanks the search results if there has been an error. We just want to carry on as normal, but without the search results if something went wrong.

File: client/lib/wpcom-undocumented/lib/undocumented.js [hunk 3]
```diff
@@ -2020,6 +2020,15 @@ Undocumented.prototype.getHelpLinks = function( searchQuery, fn ) {
 	}, fn );
 };
 
+Undocumented.prototype.getQandA = function( searchQuery, fn ) {
+	debug( 'help-contact-qanda/ searchQuery {searchQuery}' );
+
+	return this.wpcom.req.get( '/help/qanda', {
+		query: searchQuery,
+		site: config( 'happychat_support_blog' ),
```
Comment (notnownikki) [inline client/lib/wpcom-undocumented/lib/undocumented.js]: I'm not entirely sure on the debug line, I was just following the same style as used in getHelpLinks

**Comment by notnownikki** [comment]
@dmsnell I'm not familiar with any of the wpcom styles, this was my first venture into wpcom JS :)

I looked at how the help search worked for this, which does things in the same way as this branch, but if there's a preferred way, can you show me an example?

File: client/me/help/help-contact-form/index.jsx [hunk 4]
```diff
@@ -99,6 +114,17 @@ export const HelpContactForm = React.createClass( {
...
+			if ( null === error ) {
+				this.setState( { qanda: data } );
+			} else {
+				this.setState( { qanda: [] } );
+			}
+		} );
```
Comment (dmsnell) [inline client/me/help/help-contact-form/index.jsx]: if you leave out the callback you should be free to use the Promise convention, where `fn` will be `undefined` to wpcom

```js
wpcom.getQandA( query )
	.then( qanda => this.setState( { qanda } ) )
	.catch( () => this.setState( { qanda: [] } ) );
```

Comment (notnownikki) [inline client/me/help/help-contact-form/index.jsx]: Ah, sorry, for a moment there I thought you meant some wpcom specific Promise enhancements or behaviour :) Have amended this!

Comment (dmsnell) [inline client/me/help/help-contact-form/index.jsx]: was not a big deal - just wanted to make sure you knew of it. in this case it doesn't change much

**Comment by notnownikki** [comment]
Yes, I knew :) But I just went with the style that calls to getHelpLinks
used because that was where I looked to see how things were implemented.

Anyway, have changed it, I think it's more readable using then/catch

**Comment by johngodley** [comment]
LGTM :shipit:

**Comment by johngodley** [comment]
One thing, maybe change the 'fix debug call' commit message to 'Improve getQandA debug statement', or something like that? Or squash it away

**Pull request merged**